lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -69,12 +77,60 @@ LogicalResult CreateSiFiveMetadataPass::emitMemoryMetadata() {
       !dutMod ||
       instancePathCache.instanceGraph.getTopLevelNode()->getModule() == dutMod;
 
+  auto *context = circuitOp.getContext();
+  auto unknownLoc = mlir::UnknownLoc::get(context);
+  auto builderOM =
+      mlir::ImplicitLocOpBuilder::atBlockEnd(unknownLoc, moduleOp.getBody());
+
+  // Add all the properties of a memory as fields of the class.
+  // The types must match exactly with the FMemModuleOp attribute type.
+
+  StringRef paramNames[] = {"name",           "depth",        "width",
+                            "maskBits",       "readPorts",    "writePorts",
+                            "readwritePorts", "writeLatency", "readLatency"};
+  mlir::Type classFieldTypes[] = {
+      om::SymbolRefType::get(context),
+      mlir::IntegerType::get(context, 64, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned),
+      mlir::IntegerType::get(context, 32, IntegerType::Unsigned)};
+
+  om::ClassOp memorySchemaClass;
+  auto createMemoryOMSchema = [&]() {
+    // Memory metadata class.
+    memorySchemaClass =
+        builderOM.create<circt::om::ClassOp>("MemorySchema", paramNames);
+
+    // Now add all the properties of the memory as a ClassFieldOp.
+    Block *body = new Block();
+    memorySchemaClass.getRegion().push_back(body);
```

Comment:
I think you can use emplaceBlock, or something like that, rather than allocating yourself.